### ISSUE TITLE
Potentially misleading compiler error for unfinished line with Option<T>

### ISSUE THREAD

**Issue by DoumanAsh** [rust-lang/rust#25133]
Next code:

```
use std::io::{Read, stdin};

fn main() {
    stdin().bytes().next()
}
```

Gives this compile error:

```
E:\Downloads\Git\collectionScripts\rust
 [master +2 ~1 -0 !] > rustc.exe .\temp.rs
.\temp.rs:4:5: 4:27 error: mismatched types:
 expected `()`,
    found `core::option::Option<core::result::Result<u8, std::io::error::Error>>`
(expected (),
    found enum `core::option::Option`) [E0308]
.\temp.rs:4     stdin().bytes().next()
                ^~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

Which is a bit misleading as there is actually we're missing either ; or unwrap() i think?

Rust version:

```
rustc 1.0.0-beta.4 (850151a75 2015-04-30) (built 2015-05-01)
binary: rustc
commit-hash: 850151a75709f146addd30bbbf1f23d384f0b381
commit-date: 2015-04-30
build-date: 2015-05-01
host: x86_64-pc-windows-gnu
release: 1.0.0-beta.4
```

**Comment by frewsxcv** [comment]
`fn main() { }` is a function that expects nothing (`()`) to be returned. In your code block, since that statement doesn't end in a `;`, it is going to try to return the result of it to the function.

**Comment by sinistersnare** [comment]
I agree, error messages for wrong function return types should be better.

**Comment by DoumanAsh** [comment]
Oh.. i actually did not notice it at the beginning. Originally i had this problem in a bit more bigger code block.
I think it would be good to point in case of function that it is being return type of function so that it would be more clear to user why compiler is angry

**Comment by steveklabnik** [comment]
Triage: `main`'s diagnostics have a long history, https://github.com/rust-lang/rust/issues/35946 for example, and all of the `try!` stuff before that. Looks like there's a PR for this issue.

**Comment by Mark-Simulacrum** [comment]
I think the semicolon suggestion we give today is a good start, but indicating that the cause of the expected type is the function's return type would be even better.

```
error[E0308]: mismatched types
 --> test.rs:4:5
  |
4 |     stdin().bytes().next()
  |     ^^^^^^^^^^^^^^^^^^^^^^- help: did you mean to add a semicolon here?: `;`
  |     |
  |     expected (), found enum `std::option::Option`
  |
  = note: expected type `()`
             found type `std::option::Option<std::result::Result<u8, std::io::Error>>`

error: aborting due to previous error
```
